### DISCUSSION TITLE
Opting out of build pre-rendering at the Route Segment level

### DISCUSSION THREAD

**Discussion by ernestoresende** [vercel/next.js#50190] [App Router]
This is a thought exercise I've been having a conflict with since last week while working on a project.

Assume the following scenario. The initial page makes a fetch request to one of our Route Handlers, and sets a revalidation rule to something like 60 seconds, making it an ISR route segment:

```javascript
// Page Segment
export default async function Page() {
  const revalidatedData = await fetch(`http://.../api/greetings`, {
    next: { revalidate: 60 },
  });
 
  return <div>...</div>;
}
```

```typescript
// Route handler
import { NextResponse } from "next/server";

export async function GET() {
  return NextResponse.json({ message: "Greetings from Route Handler"}, {
    status: 200,
  });
}
```

You try building it locally or deploying on Vercel, and... the `fetch`  request fails, aborting the build. Thing is, this is the first deployment. So any attempts at pre-rendering the page at build time will fail because, at this point in time, the endpoint does not exist.

Sure, Sever Components let's us ask for data directly from the database or "sources other than a network request", but let's assume that, for one or more reasons, this is not an option here. We could also change the code for this page's rendering behavior to be fully dynamic on the first deploy, and then change it back to the previous state on the next deploy (that way we have a valid, functioning endpoint that will serve data if hit during the build pre-render stage). 

It works, but now we have this weird situation where you depend on the build of your previous deployment in order to fetch data on the pre-render stage (it doesn't *really* represent a problem per-see, unless the implementation of the Route Handler also changes between deploys, and the probability of that happening shifts wildly from project to project). Kind of an egg/chicken situation.

With that (pretty specific, yet valid, and I assume somewhat common) pattern, am I way too out of my league when asking if an option to opt-out of build time pre-rendering should be considered for future releases?

Back in the pages directory, this was an option for ISR when generating new paths with `getStaticPaths()` and returning `fallback: 'blocking'` . From the currently available [Pages Directory documentation](https://nextjs.org/docs/pages/building-your-application/rendering/incremental-static-regeneration):

> When a request is made to a path that hasn’t been generated, Next.js will server-render the page on the first request. Future requests will serve the static file from the cache.

While acknowledging that in this case we're talking about incremental page generation based on fetched data, at least conceptually, this can also be achieved with almost the same behavior on the App Router: by opting out of pre-render at build time, the first request will have the same behavior of exporting `export const dynamic = 'force-dynamic'`, and subsequent requests can continue to operate in normal ISR fashion (or even SSG, if revalidation isn't needed). 

**TLDR**; A way of telling Next: "*don't try fetching and pre-rendering this route segment on build, because the data it depends on is coming from somewhere (Next.js's very own Route Handler) that doesn't yet exist until this deployment is finished*".

Would love to get some insight from someone who understands the App Directory internals better then me, in order to understand if this is somehow currently possible, and if not, if it has any chance of achieving implementation on some future release.

**Comment by nabildroid** [comment]
use `generateStaticParams` and return garbage data as params 
in layout check the param if it is the garbage value then simply return empty div
